lib/src/value/color.ts

```diff
@@ -3,371 +3,1112 @@
+  /**
+   * This color's saturation in the HSL color space, between `0` and `100`.
+   *
+   * @deprecated Use {@link channel} instead.
+   */
   get saturation(): number {
-    if (this.saturationInternal === undefined) {
-      this.rgbToHsl();
-    }
-    return this.saturationInternal!;
+    emitColor4ApiGetterDeprecation('saturation');
+    return NaNtoZero(this.color.hsl.saturation);
   }
 
-  /** `this`'s hue value. */
+  /**
+   * This color's lightness in the HSL color space, between `0` and `100`.
+   *
+   * @deprecated Use {@link channel} instead.
+   */
   get lightness(): number {
-    if (this.lightnessInternal === undefined) {
-      this.rgbToHsl();
-    }
-    return this.lightnessInternal!;
+    emitColor4ApiGetterDeprecation('lightness');
+    return NaNtoZero(this.color.hsl.lightness);
   }
 
-  /** `this`'s whiteness value. */
+  /**
+   * This color's whiteness in the HWB color space, between `0` and `100`.
+   *
+   * @deprecated Use {@link channel} instead.
+   */
   get whiteness(): number {
-    // Because HWB is (currently) used much less frequently than HSL or RGB, we
-    // don't cache its values because we expect the memory overhead of doing so
-    // to outweigh the cost of recalculating it on access.
-    return (Math.min(this.red, this.green, this.blue) / 255) * 100;
+    emitColor4ApiGetterDeprecation('whiteness');
+    return NaNtoZero(this.color.hwb.whiteness);
   }
 
-  /** `this`'s blackness value. */
+  /**
+   * This color's blackness in the HWB color space, between `0` and `100`.
+   *
+   * @deprecated Use {@link channel} instead.
+   */
   get blackness(): number {
-    // Because HWB is (currently) used much less frequently than HSL or RGB, we
-    // don't cache its values because we expect the memory overhead of doing so
-    // to outweigh the cost of recalculating it on access.
-    return 100 - (Math.max(this.red, this.green, this.blue) / 255) * 100;
+    emitColor4ApiGetterDeprecation('blackness');
+    return NaNtoZero(this.color.hwb.blackness);
   }
 
-  /** `this`'s alpha channel. */
-  get alpha(): number {
-    return this.alphaInternal;
+  assertColor(): SassColor {
+    return this;
+  }
+
+  // Internal helper which mutates the current color object.
+  _toSpaceInternal(space: KnownColorSpace): void {
+    this.isRgb = space === 'rgb';
+    this.color = this.color.to(encodeSpaceForColorJs(space) as string);
   }
 
   /**
-   * Whether `this` has already calculated the HSL components for the color.
-   *
-   * This is an internal property that's not an official part of Sass's JS API,
-   * and may be broken at any time.
+   * Returns a new color that's the result of converting this color to the
+   * specified `space`.
    */
-  get hasCalculatedHsl(): boolean {
-    return !!this.hueInternal;
+  toSpace(space: KnownColorSpace): SassColor {
+    if (space === this.space) return this;
+    const color = this.clone();
+    color._toSpaceInternal(space);
+    return color;
```

Comment:
I prefer to avoid making value objects mutable if possible, even just internally—it opens up a huge can of worms when trying to track down subtle bugs. Can we add a secret constructor overload that just takes a color object instead?

Reply:
Reworked to use a constructor in https://github.com/sass/embedded-host-node/pull/259/commits/18ab10ee84ce3b3321b8a8b98f52c7fb79650653